### ISSUE TITLE
Duel System for Choosing heroes

### ISSUE THREAD

**Issue by mmdts** [darklordabc/chc2issues#138]
#### Current duel algorithm:

> ```
> 1. get all teams
> 2. if there are at least 5 alive (non-eliminated) teams, discard the 2 teams that fought in the previous duel
> 3. discard all teams with 2+ duel participations MORE than the team with lowest duel participations
> 4. if at least 2 teams remain after the steps above, set up a duel with any of them (picked randomly)
> 5. if there aren't enough teams after step 3, re-add the teams discarded there
> 6. match any 2 of the remaining teams randomly
> ```

___

#### Suggested duel algorithm V1:

> I devised a better algorithm for dueling that:
> ```
> 1) Makes it so that players that are missing out on duels are MUCH more likely to be picked in duels, like the current system.
> 2) Is mathematically stable and does not make the same two players always duel each other, unlike the current system.
> ```
> 
> The dueling algorithm goes as follows:
> ```
> 0) The chance of dueling on the first round is 0.
> 1) On round start, a random number is rolled to decide if a duel takes place. The chance of a duel taking place is the amount of living players/teams divided by the total player/team count. This makes duels 100% common till a player/team dies, with them being much rarer at the end (20% common when only 2 players/teams are alive).
> If the duel takes place:
> 2) LIVING player/team weights are used for weighted random sampling of which player/team will duel, players/teams with higher weights are much more likely to participate in duels. [*]
> 3) The first player/team is removed from the list of living players/teams (to prevent them from duelling themselves) and a second player/team is selected through weighted random sampling.
> 4) New player weights are calculated as follows:
> player_weight_for_this_player = the_amount_of_duels_by_the_highest_dueling_player - the_amount_of_duels_by_this_player * FULL + random(0~1) * BIAS [**]
> 5) If only 2 players are remaining and a duel took place this round, the chance of a duel on the next round becomes 0.
> ```
> 
> [*] (https://en.wikipedia.org/wiki/Reservoir_sampling#Weighted_random_sampling)
>
> [**] Recommended values are FULL = 100, BIAS = 40, higher BIAS:FULL ratio means higher likelihood for a given player to be selected even though they aren't the least dueling player.
> 
> This system is based on my attempts to make a fair dueling system that solves the shortcomings of the current one, starting from completely uniformly random dueling to this system in 4 steps.

Simulation for this algorithm (in Javascript) is present at: https://gist.github.com/mmdts/e3a66a7100ac700d2562963efa239b73#file-3i_biased_weighted_random_duels-js

The final result is the file `3i_biased_weighted_random_duels.js` which can be run as follows:
```bash
$ node 3i_biased_weighted_random_duels.js > output.txt
```
Other intermediate thoughts and duel systems can be found in the other files, explaining how I reached this specific algorithm.

The most important part of this simulation is the weighted random number function,
```javascript
function weighted_random(weights) {
  let sum = weights.reduce((acc, o) => acc + o);
  let rnd = ~~(Math.random() * sum);
  for (let i = 0; i < weights.length; i++) {
    if (rnd - weights[i] < 0) return i;
    rnd -= weights[i];
  }
  return ~~(Math.random() * weights.length);
}
```
used to determine duelists:
```javascript
weighted_random_result = weighted_random(living_player_weights);
duelist_1 = living_players.splice(weighted_random_result, 1)[0];
```
___

#### Firetoard remarks on V1:

Firetoad said the duel system looks very fair, and wanted to implement it:

> hey, just want to say I went through your last dump, I think it works pretty nicely, gonna try implementing it with the only change being deterministic intervals between duels 

 However, he complained that the duel system does not preserve the current duel frequency. He wanted to implement my system, but still pick heroes deterministically based on the number of players alive, which currently goes:

> for ffa as an example, there's a duel every round while there are 9-10 players, then one every other round for 6-8, one every 3 rounds for 4-5, then finally one duel every 4 rounds for 3 or less players

I personally believe the current duel frequency is broken for reasons that it might duel the same two people twice in a row if someone dies in the middle, which is why I think studying the current duel frequency's cumulative distribution, which is based on the cumulative distribution of the player elimination chance random variable, and then devising a randomized duel system that preserves the same exact duel frequency would be worth it some time in the future.

___

#### Suggested duel algorithm V2:

This is the same as V1, but preserves the static duel frequency at Firetoad's request (affects step 1).

```
0) The chance of dueling on the first round is 0.
1) On round start, the game checks how many players are alive to decide if a duel takes place. There's a duel every round while there are 9-10 players, then one every other round for 6-8, one every 3 rounds for 4-5, then finally one duel every 4 rounds for 3 or less players.
If the duel takes place:
2) LIVING player/team weights are used for weighted random sampling of which player/team will duel, players/teams with higher weights are much more likely to participate in duels.
3) The first player/team is removed from the list of living players/teams (to prevent them from dueling themselves) and a second player/team is selected through weighted random sampling.
4) New player weights are calculated as follows:
player_weight_for_this_player = the_amount_of_duels_by_the_highest_dueling_player - the_amount_of_duels_by_this_player * FULL + random(0~1) * BIAS
5) If only 2 players are remaining and a duel took place this round, the chance of a duel on the next round becomes 0.

FULL = 100, BIAS = 40
```
___

I don't mind offering any help I can offer in order to fix the duel system, as it's really been bugging me.

@AustraliaIsMyCity

**Comment by Draze22** [comment]
Didn't read the entire thing but your solution looks like a good one. However if the devs want an easy to implement system that is similar to the current one but stops teams from dueling the team repeatedly you just need to remove the
![image](https://user-images.githubusercontent.com/63279022/80231728-2197c680-8697-11ea-9737-6bf74fc45b5d.png)
from the code and then add another bit I'll talk about later. 

The most vivid example of why this bit of code is problematic is when you have 4 teams remaining. 2 teams duel each other and are then excluded from the next duel leaving the only option to make the other 2 team duel each other creating an infinite loop of the same 2 sets of teams dueling each other until one team is eliminated.

What you will want to add instead is a bit of code that stops the two team that just faced each other from facing each other again. In other words let one team be in 2 duels in a row but don't let the same 2 teams be in 2 duels in a row.

**Comment by mmdts** [comment]
You don't solve a problem that's caused by wacky mathematics by removing wacky bits and adding others.
You fix the mathematics.

It's just the same way [PRD](https://liquipedia.net/dota2/Pseudo_Random_Distribution) is a fix for independent uniform random distribution causing long proc droughts or streaks. Imagine if dota2 devs decided "hey, lets add a rule that makes sure it never procs thrice in a row!" or "hey lets add a rule that makes sure it doesn't roll over 7 times without proccing!"

That's be correct in concept, but would be far from the correct solution the issue, compared to *correct issue-fixing mathematics* like PRD.

**Comment by Draze22** [comment]
Never said my solution was better than yours, just that it was better than the current one and easy to implement. I imagine it would take the devs a while to code yours but it should only take ~10 minutes to do mine.

To be honest I would be fine if 2 teams were able to fight each other 2 duels in a row (or more) and just remove the section I suggested removing completely and not add my suggested improvement.

You haven't actually given a reason why my solution would cause more problems than the current one where as I have provided reasons why it would be better than the current one until yours can be fully understood and implemented by the devs. In other words use my solution as a hotfix and yours as the long term duel algorithm (if they think your solution is ideal).
